tools/clang/lib/Sema/SemaOverload.cpp

```diff
@@ -4967,8 +4967,7 @@ InitCallParamConversions(Sema &S, const FunctionProtoType *Proto,
     Expr *OutFrom = DeclRefExpr::Create(
         S.getASTContext(), NestedNameSpecifierLoc(), SourceLocation(), Param,
         true, Param->getLocation(), ParamType.getNonReferenceType(), VK_RValue, nullptr);
-    OutConversion = TryCopyInitialization(
-        S, OutFrom, Arg->getType(), SuppressUserConversions,
+    OutConversion = TryCopyInitialization(S, OutFrom, ParamType, SuppressUserConversions,
```

Comment:
I’m not sure this change is correct. This alters the AST so that instead of converting the return value from Parent->Child, it converts Parent->Parent.

Reply:
It also looks like we don't support `HLSLDerivedToBase` truncation for `inout` or `out` on other parameters (https://godbolt.org/z/aqaKK1qTd), so we don't need to for `this` either.

Reply:
Reverted.

Reply:
Isn't it somewhat strange to have different cast behavior depending on whether the var is `in` only or `inout`/ `out`?

Reply:
Yeah, it is strange.
Do you want to always require cast or always implicit cast for case like this?

Reply:
This code is wonky, but I think correct.

For `in` and `inout` parameters we copy initialize a new object from the argument type to the parameter type.

Then for `out` and `inout` parameters we copy initialize from the parameter to the argument type.

The trick here is that `InConversion` and `OutConversion` are two different casts, and they can both be set.

This should correctly model the semantics that `in` and `inout` parameters need to be casted going in, and that `out` and `inout` need to be casted coming out.

Reply:
What's the point to cast to bigger type when the copy has to be truncated?

If cast to parameter type, the copy will be on smaller type which is just normal copy.

Reply:
It isn't really a cast. It is copy-initializing.

On `in` and `inout` parameters the argument (value being passed in) is copy-initialized to a temporary of the parameter type, which is passed into the function. For `out` parameters the temporary is uninitialized.
On `out` and `inout` parameters after the call, the temporary value of the parameter-type is copy-initialized back to the argument variable.

So the type conversion go from argument -> parameter on input and from parameter -> argument on output, but these are copy-initializations not casts.